File: VirtualLab_Python/SynDiag_Python/diagnostics/Tokalab/Diag_ThomsonScattering.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun  3 17:48:05 2025

@author: ricca
"""

import numpy as np
from scipy.interpolate import griddata
import matplotlib.pyplot as plt

class Diag_ThomsonScattering:
    def __init__(self):
        self.R = None  # Horizontal coordinate
        self.Z = None  # Vertical coordinate
        self.ne = None  # Measured electron density
        self.sigma_ne = None # Uncertainty of the measured electron density
        self.Te = None  # Measured electron temperature
        self.sigma_Te = None # Uncertainty of the measured electron temperature
        self.unit_ne = None  # Unit for ne
        self.unit_Te = None  # Unit for Te
        self.config = {}  # Configuration dictionary
        self.ideal = {}  # Ideal (noise-free) values

    def measure(self, equi):
                
        points = np.column_stack((equi.geo.grid.Rg.flatten(), equi.geo.grid.Zg.flatten()))

        # Interpolazione con griddata (equiv. interp2 Matlab)
        ne = griddata(points, equi.ne.flatten(), (self.R, self.Z), method='linear')
        Te = griddata(points, equi.Te.flatten(), (self.R, self.Z), method='linear')
        
        self.ideal["ne"] = ne
        self.ideal["Te"] = Te

        noise_abs_ne = np.random.normal(0, self.config.get("ne_noise_random_absolute_intensity", 0), size=np.shape(ne))
        noise_abs_Te = np.random.normal(0, self.config.get("Te_noise_random_absolute_intensity", 0), size=np.shape(Te))

        noise_prop_ne = np.random.normal(0, np.abs(ne) * self.config.get("ne_noise_random_proportional_intensity", 0))
        noise_prop_Te = np.random.normal(0, np.abs(Te) * self.config.get("Te_noise_random_proportional_intensity", 0))

        self.ne = ne + noise_abs_ne + noise_prop_ne
        self.Te = Te + noise_abs_Te + noise_prop_Te

        self.sigma_ne = np.sqrt(self.config.get("ne_noise_random_absolute_intensity", 0)**2 +
                                np.abs(ne) * self.config.get("ne_noise_random_proportional_intensity", 0)**2)
        self.sigma_Te = np.sqrt(self.config.get("Te_noise_random_absolute_intensity", 0)**2 +
                                (np.abs(Te) * self.config.get("Te_noise_random_proportional_intensity", 0))**2)
        
        self.unit_ne = "m^{-3}"
        self.unit_Te = "eV"

    def upload(self, configuration=1):
        if configuration == 1:
            self.config["configuration"] = 1

            self.R = np.linspace(6, 8.4, 60)
            self.Z = np.linspace(0, 0.5, 60)

            self.config["ne_noise_random_absolute_intensity"] = 0
            self.config["Te_noise_random_absolute_intensity"] = 0
            self.config["ne_noise_random_proportional_intensity"] = 0
            self.config["Te_noise_random_proportional_intensity"] = 0
            
    #### Plotting function
    def plot_geo(self):
        plt.plot(self.R, self.Z, '.', markersize=16)
        plt.grid(visible=True, which='both')
        plt.xlabel("R")
        plt.ylabel("Z")
        plt.show()
        
    def plot_ne_meas(self):
        plt.plot(self.ne, '.', markersize=16)
        plt.grid(visible=True, which='both')
        plt.xlabel("#")
        plt.ylabel("n_e [m$^{-3}$]")
        plt.show()
        
    def plot_Te_meas(self):
        plt.plot(self.Te, '.', markersize=16)
        plt.grid(visible=True, which='both')
        plt.xlabel("#")
        plt.ylabel("T_e [eV]")
        plt.show()
        
    def plot_StandAlone(self):
        fig, axes = plt.subplots(1, 2, figsize=(12, 5))
    
        # Te plot
        axes[0].plot(self.ideal["Te"], '.b', markersize=16)
        axes[0].plot(self.Te, 'or', linewidth=1.2)
        axes[0].grid(visible=True, which='both')
        axes[0].set_xlabel("#")
        axes[0].set_ylabel(r"$T_e$ [eV]")
    
        # ne plot
        axes[1].plot(self.ideal["ne"], '.b', markersize=16)
        axes[1].plot(self.ne, 'or', linewidth=1.2)
        axes[1].grid(visible=True, which='both')
        axes[1].set_xlabel("#")
        axes[1].set_ylabel(r"$n_e$ [m^{-3}]")
    
        plt.tight_layout()
        plt.show()    
        
        
        
        
        
        
        
        

File: VirtualLab_MatLab/SynDiag_MATLAB/diagnostics/Diag_ThomsonScattering.m
%% tokalab diagnostics

classdef Diag_ThomsonScattering

    properties

        R % Horixontal coordinate
        Z % Vertical coordinate

        ne % Measured Electron Density
        Te % Measured Electron Temperature

        sigma_ne % Associated Uncertainty to Measured Electron Density
        sigma_Te % Associated Uncertainty to Measured Electron Temperature

        unit_Te % Unit Measure Electron Temperature
        unit_ne % Unit Measure Electron Density

        config % contains the various information such as noise, etc.

        ideal % contains the measurements without the noise

    end

    methods

        function obj = measure(obj,equi)

            R_equi = equi.geo.grid.Rg;
            Z_equi = equi.geo.grid.Zg;

            ne_equi = equi.ne;
            Te_equi = equi.Te;

            obj.ideal.ne = interp2(R_equi,Z_equi,ne_equi,obj.R,obj.Z);
            obj.ideal.Te = interp2(R_equi,Z_equi,Te_equi,obj.R,obj.Z);

            % noise absolute
            noise_abs_ne = normrnd(0,obj.config.ne_noise_random_absolute_intensity,size(obj.ideal.ne));
            noise_abs_Te = normrnd(0,obj.config.Te_noise_random_absolute_intensity,size(obj.ideal.Te));

            % noise proportional
            noise_prop_ne = normrnd(0,abs(obj.ideal.ne).*obj.config.ne_noise_random_proportional_intensity);
            noise_prop_Te = normrnd(0,abs(obj.ideal.Te).*obj.config.Te_noise_random_proportional_intensity);

            % real measurement
            obj.ne = obj.ideal.ne + noise_abs_ne + noise_prop_ne;
            obj.Te = obj.ideal.Te + noise_abs_Te + noise_prop_Te;

            % associated uncertainties
            obj.sigma_ne = sqrt(obj.config.ne_noise_random_absolute_intensity.^2 + ...
                (abs(obj.ideal.ne).*obj.config.ne_noise_random_proportional_intensity).^2);
            obj.sigma_Te = sqrt(obj.config.Te_noise_random_absolute_intensity.^2 + ...
                (abs(obj.ideal.Te).*obj.config.Te_noise_random_proportional_intensity).^2);

            obj.unit_Te = "eV";
            obj.unit_ne = "m^{-3}";

        end

        %% Functions
        function obj = Upload(obj,configuration,machine)

            % default configuration
            if nargin<2
                configuration = 1;
                machine = "TokaLab";
            elseif nargin < 3
                machine = "TokaLab";
            end

            %%%%%%%%%%%% TokaLab Configuration %%%%%%%%%%%%%%%%%%%%%%%%%%%%

            if machine == "TokaLab"

                if configuration == 1

                    obj.config.configuration = 1;

                    obj.R = linspace(6,8.4,60);
                    obj.Z = linspace(0,0.5,60);

                    obj.config.ne_noise_random_absolute_intensity = 0;
                    obj.config.Te_noise_random_absolute_intensity = 0;

                    obj.config.ne_noise_random_proportional_intensity = 0;
                    obj.config.Te_noise_random_proportional_intensity = 0;

                end
            end

            %%%%%%%%%%%% JET-like Configuration %%%%%%%%%%%%%%%%%%%%%%%%%%%%

            if machine == "JET-like"

                if configuration == 1

                    obj.config.configuration = 1;

                    obj.R = linspace(2.98,3.88,60);
                    obj.Z = linspace(0.06,0.11,60);

                    obj.config.ne_noise_random_absolute_intensity = 0;
                    obj.config.Te_noise_random_absolute_intensity = 0;

                    obj.config.ne_noise_random_proportional_intensity = 0;
                    obj.config.Te_noise_random_proportional_intensity = 0;

                end
            end

            %%%%%%%%%%%% DTT-like Configuration %%%%%%%%%%%%%%%%%%%%%%%%%%%%

            if machine == "DTT-like"

                if configuration == 1

                    obj.config.configuration = 1;

                    obj.R = linspace(6,8.4,60);
                    obj.Z = linspace(0,0.5,60);

                    obj.config.ne_noise_random_absolute_intensity = 0;
                    obj.config.Te_noise_random_absolute_intensity = 0;

                    obj.config.ne_noise_random_proportional_intensity = 0;
                    obj.config.Te_noise_random_proportional_intensity = 0;

                end
            end
        end

        %% Plotting Functions

        function plot_geo(obj)

            plot(obj.R,obj.Z,'.','MarkerSize',16)
            grid on
            grid minor
            xlabel("R")
            ylabel("Z")

        end

        function plot_Ne_meas(obj)

            plot(obj.ne,'.','MarkerSize',16)
            grid on
            grid minor
            xlabel("#")
            ylabel("N_e [m^{-3}]")

        end

        function plot_Te_meas(obj)

            plot(obj.Te,'.','MarkerSize',16)
            grid on
            grid minor
            xlabel("#")
            ylabel("T_e [m^{-3}]")

        end

        function plot_StandAlone(obj)

            subplot(1,2,1)
            hold off
            plot(obj.ideal.Te,'.b','MarkerSize',16)
            hold on
            plot(obj.Te,'or','LineWidth',1.2)
            grid on
            grid minor
            xlabel("#")
            ylabel("T_e [eV]")

            subplot(1,2,2)
            hold off
            plot(obj.ideal.ne,'.b','MarkerSize',16)
            hold on
            plot(obj.ne,'or','LineWidth',1.2)
            grid on
            grid minor
            xlabel("#")
            ylabel("n_e [m^{-3}]")

        end

    end

end




